Source/JavaScriptCore/runtime/JSDateMath.cpp

```diff
@@ -185,103 +189,181 @@ LocalTimeOffset DateCache::calculateLocalTimeOffset(double millisecondsFromEpoch
+LocalTimeOffsetCache* DateCache::DSTCache::leastRecentlyUsed(LocalTimeOffsetCache* exclude)
+{
+    LocalTimeOffsetCache* result = nullptr;
+    for (auto& cache : m_entries) {
+        if (&cache == exclude)
+            continue;
+        if (!result) {
+            result = &cache;
+            continue;
+        }
+        if (result->epoch > cache.epoch)
+            result = &cache;
+    }
+    *result = LocalTimeOffsetCache { };
+    return result;
+}
+
+std::tuple<LocalTimeOffsetCache*, LocalTimeOffsetCache*> DateCache::DSTCache::probe(int64_t millisecondsFromEpoch)
 {
-    LocalTimeOffsetCache& cache = inputTimeType == WTF::LocalTime ? m_localTimeOffsetCache : m_utcTimeOffsetCache;
+    LocalTimeOffsetCache* before = nullptr;
+    LocalTimeOffsetCache* after = nullptr;
+    for (auto& cache : m_entries) {
+        if (cache.start <= millisecondsFromEpoch) {
+            if (!before || before->start < cache.start)
+                before = &cache;
+        } else if (millisecondsFromEpoch < cache.end) {
+            if (!after || after->end > cache.end)
+                after = &cache;
+        }
+    }
+
+    if (!before) {
+        if (m_before->isEmpty())
+            before = m_before;
+        else
+            before = leastRecentlyUsed(after);
+    }
+    if (!after) {
+        if (m_after->isEmpty() && before != m_after)
+            after = m_after;
+        else
+            after = leastRecentlyUsed(before);
+    }
+
+    m_before = before;
+    m_after = after;
+    return std::tuple { before, after };
+}
+
+void DateCache::DSTCache::extendTheAfterCache(int64_t millisecondsFromEpoch, LocalTimeOffset offset)
+{
+    if (m_after->offset == offset && m_after->start - defaultDSTDeltaInMilliseconds <= millisecondsFromEpoch && millisecondsFromEpoch <= m_after->end) {
+        // Extend the m_after cache.
+        m_after->start = millisecondsFromEpoch;
+        dataLogLnIf(JSDateMathInternal::verbose, "Cache extended1 from ", millisecondsFromEpoch, " to ", m_after->end, " ", offset.offset, " ", offset.isDST);
+    } else {
+        // The m_after cache is either invalid or starts too late.
+        if (!m_after->isEmpty()) {
+            // If the m_after cache is valid, replace it with a new cache.
+            m_after = leastRecentlyUsed(m_before);
+        }
+        m_after->start = millisecondsFromEpoch;
+        m_after->end = millisecondsFromEpoch;
+        m_after->offset = offset;
+        m_after->epoch = bumpEpoch();
+        dataLogLnIf(JSDateMathInternal::verbose, "Cache miss, recompute2 ", millisecondsFromEpoch, " ", offset.offset, " ", offset.isDST);
+    }
+}
+
+LocalTimeOffset DateCache::DSTCache::localTimeOffset(DateCache& dateCache, int64_t millisecondsFromEpoch, WTF::TimeType inputTimeType)
+{
+    if (millisecondsFromEpoch >= -WTF::Int64Milliseconds::maxECMAScriptTime && millisecondsFromEpoch <= WTF::Int64Milliseconds::maxECMAScriptTime) {
```

Comment:
nit: Maybe this should be `millisecondsFromEpoch >= -minECMAScriptTime` since we have `minECMAScriptTime` defined as -8.64E15 (i.e. -maxECMAScriptTime) in JSDateMath.h.

Should the definition of `minECMAScriptTime` should be changed to be `-WTF::maxECMAScriptTime` instead of using the literal?  I'm not sure which is best.

OTOH, v8 uses 0 for minECMAScriptTime instead.  Do you know why they don't use `-WTF::maxECMAScriptTime`?

Reply:
> OTOH, v8 uses 0 for minECMAScriptTime instead. Do you know why they don't use -WTF::maxECMAScriptTime?

V8 is conservative here to support non ICU TimeZones (so avoiding negative numbers). But we do not need to do it.
And rather, we should support negative one since it makes Gregorian / Julius calendar difference unhandled.

Reply:
I defined `WTF:: Int64Milliseconds ::minECMAScriptTime`.

Reply:
Sounds good.